Enabled CMake compilation of IVAN for MinGW

stoperro opened the pull request:
Some fixes needed to compile with MinGW.
Updated install instruction.

stoperro committed:
Instructions on how to compile with MinGW.

stoperro committed:
Now when using MinGW "install" target will work.

emlai commented:
I think the `Main` functions should stay as `Main`. The real `main` is in `femain.cpp`, which calls `Main`.

emillaine commented on FeLib/Source/whandler.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "error.h"
 #include "bitmap.h"
 #include "festring.h"
+#include <stdio.h>
```
`stdio` is already included on line 13 :)

stoperro commented:
Regarding Main function - the problem is, since it's FeLib who includes SDL, order of libraries for linking is:

../FeLib/libFeLib.a -lmingw32 D:/Apps/SDL2-devel-2.0.4-mingw/SDL2-2.0.4/i686-w64-mingw32/lib/libSDL2main.a D:/Apps/SDL2-devel-2.0.4-mingw/SDL2-2.0.4/i686-w64-mingw32/lib/libSDL2.dll.a ...

Same order is on all platforms, but this is issue only on MinGW since:
- on Windows (and some other platforms) SDL changes regular "main" functions to "SDL_main" - see STL_main.h
- GCC compilers have a bit different linking principles than VS, so it's visible only there.

This is causes problem in linking because FeLib defines first symbol "SDL_main" (changed from "main"), before it's requested by SDL which is linked later. On GCC this means symbol is discarded before it gets to SDL linking. This doesn't happen on Linux, because symbol is "main" which is requested by compiler always.

One simple solution for that is to delete femain.cpp, since it's usability is now very small (catching unhandled exceptions could be moved to regular main.cpp). I'm not sure if this if good solution in relation to IVAN architecture. If it's ok, I can propose needed changes.

emlai commented:
Thanks for the explanation. I think it would be best to remove femain.cpp altogether to simplify things. After all, the additional code there's has been commented out for a long time. I'll make a commit for that and add it to this pull request.

emillaine committed:
Remove femain.cpp

This is made to simplify linking on MinGW. For more info, see:
https://github.com/emlai/ivan/pull/3#issuecomment-246200091

emlai commented:
Removed femain.cpp and it works good on MinGW. Will test with VS and Mac to see that it didn't break anything.

By the way, I get some weird errors when compiling on MinGW with `-std=c++11` and changing that to `-std=gnu++11` makes them go away? Do you know a proper solution to this? Or should we specify in the MinGW instructions to add `-D CMAKE_CXX_FLAGS='-std=gnu++11'` when invoking CMake?

stoperro commented:
Yes, it still works on Visual Studio.

Note there is a crash on shutdown, caused by thise lines:
hscore.h:
...
#ifdef UNIX
#define HIGH_SCORE_FILENAME LOCAL_STATE_DIR "/ivan-highscore.scores"
#endif

#if defined(WIN32) || defined(__DJGPP__)
#define HIGH_SCORE_FILENAME CONST_S("HScore.dat")
#endif


In save.cpp:
outputfile::outputfile(cfestring& FileName, truth AbortOnErr)
: FileName(FileName)
{
  // Make sure the directory exists first.
  festring DirectoryPath = FileName;
  DirectoryPath.Resize(FileName.FindLast('/'));


 	ivan.exe!festring::Resize(unsigned long N, char C) Line 243	C++
 	ivan.exe!outputfile::outputfile(const festring & FileName, bool AbortOnErr) Line 32	C++
>	ivan.exe!highscore::Save(const festring & File) Line 103	C++


Since on Windows filename is "HScore.dat", not "/ivan-highscore.scores", it has no '/' character and FindLast returns "NPos", which is (unsigned)-1.

stoperro commented:
Regarding MinGW errors, I'm not MinGW expert (I've switched on Windows to VS a while ago), but what are they?

I have no errors now on MinGW, my version:
>g++ --version
g++ (tdm-1) 4.9.2


Btw. good to hear it works on so many platforms now.

emillaine committed:
Fix crash when passing "HScore.dat" to outputfile ctor

emlai commented:
Ah that crash, well spotted! Added commit 3216ad6001a125263187803cf746b9ea98060ece to not create any directories if the filename is, well, just a filename, instead of a full path.

emlai commented:
My MinGW g++ version 5.3.0. The output looks like this with `-std=c++11`:

```
In file included from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\postypes.h:40:0,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:40,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\string:40,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\stdexcept:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\array:38,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\tuple:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\stl_map.h:63,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\map:61,
                 from C:/MinGW/ivan/FeLib/Include/save.h:21,
                 from C:\MinGW\ivan\FeLib\Source\bitmap.cpp:18:
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:146:11: error: '::fwide' has not been declared
   using ::fwide;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:153:11: error: '::mbsinit' has not been declared
   using ::mbsinit;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:177:11: error: '::wcscat' has not been declared
   using ::wcscat;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:178:11: error: '::wcscmp' has not been declared
   using ::wcscmp;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:179:11: error: '::wcscoll' has not been declared
   using ::wcscoll;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:180:11: error: '::wcscpy' has not been declared
   using ::wcscpy;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:181:11: error: '::wcscspn' has not been declared
   using ::wcscspn;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:183:11: error: '::wcslen' has not been declared
   using ::wcslen;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:184:11: error: '::wcsncat' has not been declared
   using ::wcsncat;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:185:11: error: '::wcsncmp' has not been declared
   using ::wcsncmp;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:186:11: error: '::wcsncpy' has not been declared
   using ::wcsncpy;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:188:11: error: '::wcsspn' has not been declared
   using ::wcsspn;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:193:11: error: '::wcstok' has not been declared
   using ::wcstok;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:196:11: error: '::wcsxfrm' has not been declared
   using ::wcsxfrm;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:198:11: error: '::wmemcmp' has not been declared
   using ::wmemcmp;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:199:11: error: '::wmemcpy' has not been declared
   using ::wmemcpy;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:200:11: error: '::wmemmove' has not been declared
   using ::wmemmove;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:201:11: error: '::wmemset' has not been declared
   using ::wmemset;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:204:11: error: '::wcschr' has not been declared
   using ::wcschr;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:205:11: error: '::wcspbrk' has not been declared
   using ::wcspbrk;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:206:11: error: '::wcsrchr' has not been declared
   using ::wcsrchr;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:207:11: error: '::wcsstr' has not been declared
   using ::wcsstr;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:208:11: error: '::wmemchr' has not been declared
   using ::wmemchr;
           ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar: In function 'wchar_t* std::wcschr(wchar_t*, wchar
_t)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:213:55: error: invalid conversion from 'const wcha
r_t*' to 'wchar_t*' [-fpermissive]
   { return wcschr(const_cast<const wchar_t*>(__p), __c); }
                                                       ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:212:3: note:   initializing argument 1 of 'wchar_t
* std::wcschr(wchar_t*, wchar_t)'
   wcschr(wchar_t* __p, wchar_t __c)
   ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar: In function 'wchar_t* std::wcspbrk(wchar_t*, cons
t wchar_t*)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:217:58: error: invalid conversion from 'const wcha
r_t*' to 'wchar_t*' [-fpermissive]
   { return wcspbrk(const_cast<const wchar_t*>(__s1), __s2); }
                                                          ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:216:3: note:   initializing argument 1 of 'wchar_t
* std::wcspbrk(wchar_t*, const wchar_t*)'
   wcspbrk(wchar_t* __s1, const wchar_t* __s2)
   ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar: In function 'wchar_t* std::wcsrchr(wchar_t*, wcha
r_t)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:221:56: error: invalid conversion from 'const wcha
r_t*' to 'wchar_t*' [-fpermissive]
   { return wcsrchr(const_cast<const wchar_t*>(__p), __c); }
                                                        ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:220:3: note:   initializing argument 1 of 'wchar_t
* std::wcsrchr(wchar_t*, wchar_t)'
   wcsrchr(wchar_t* __p, wchar_t __c)
   ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar: In function 'wchar_t* std::wcsstr(wchar_t*, const
 wchar_t*)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:225:57: error: invalid conversion from 'const wcha
r_t*' to 'wchar_t*' [-fpermissive]
   { return wcsstr(const_cast<const wchar_t*>(__s1), __s2); }
                                                         ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:224:3: note:   initializing argument 1 of 'wchar_t
* std::wcsstr(wchar_t*, const wchar_t*)'
   wcsstr(wchar_t* __s1, const wchar_t* __s2)
   ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar: In function 'wchar_t* std::wmemchr(wchar_t*, wcha
r_t, std::size_t)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:229:61: error: invalid conversion from 'const wcha
r_t*' to 'wchar_t*' [-fpermissive]
   { return wmemchr(const_cast<const wchar_t*>(__p), __c, __n); }
                                                             ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:228:3: note:   initializing argument 1 of 'wchar_t
* std::wmemchr(wchar_t*, wchar_t, std::size_t)'
   wmemchr(wchar_t* __p, wchar_t __c, size_t __n)
   ^
In file included from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\string:40:0,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\stdexcept:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\array:38,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\tuple:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\stl_map.h:63,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\map:61,
                 from C:/MinGW/ivan/FeLib/Include/save.h:21,
                 from C:\MinGW\ivan\FeLib\Source\bitmap.cpp:18:
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h: In static member function 'static int
 std::char_traits<wchar_t>::compare(const char_type*, const char_type*, std::size_t)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:353:32: error: 'wmemcmp' was not decla
red in this scope
  return wmemcmp(__s1, __s2, __n);
                                ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h: In static member function 'static std
::size_t std::char_traits<wchar_t>::length(const char_type*)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:358:26: error: 'wcslen' was not declar
ed in this scope
       { return wcslen(__s); }
                          ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h: In static member function 'static con
st char_type* std::char_traits<wchar_t>::find(const char_type*, std::size_t, const char_type&)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:365:30: error: invalid conversion from
 'const char_type* {aka const wchar_t*}' to 'wchar_t*' [-fpermissive]
  return wmemchr(__s, __a, __n);
                              ^
In file included from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\postypes.h:40:0,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:40,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\string:40,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\stdexcept:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\array:38,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\tuple:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\stl_map.h:63,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\map:61,
                 from C:/MinGW/ivan/FeLib/Include/save.h:21,
                 from C:\MinGW\ivan\FeLib\Source\bitmap.cpp:18:
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\cwchar:228:3: note:   initializing argument 1 of 'wchar_t
* std::wmemchr(wchar_t*, wchar_t, std::size_t)'
   wmemchr(wchar_t* __p, wchar_t __c, size_t __n)
   ^
In file included from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\string:40:0,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\stdexcept:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\array:38,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\tuple:39,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\stl_map.h:63,
                 from c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\map:61,
                 from C:/MinGW/ivan/FeLib/Include/save.h:21,
                 from C:\MinGW\ivan\FeLib\Source\bitmap.cpp:18:
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h: In static member function 'static std
::char_traits<wchar_t>::char_type* std::char_traits<wchar_t>::move(std::char_traits<wchar_t>::char_t
ype*, const char_type*, std::size_t)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:373:33: error: 'wmemmove' was not decl
ared in this scope
  return wmemmove(__s1, __s2, __n);
                                 ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h: In static member function 'static std
::char_traits<wchar_t>::char_type* std::char_traits<wchar_t>::copy(std::char_traits<wchar_t>::char_t
ype*, const char_type*, std::size_t)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:381:32: error: 'wmemcpy' was not decla
red in this scope
  return wmemcpy(__s1, __s2, __n);
                                ^
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h: In static member function 'static std
::char_traits<wchar_t>::char_type* std::char_traits<wchar_t>::assign(std::char_traits<wchar_t>::char
_type*, std::size_t, std::char_traits<wchar_t>::char_type)':
c:\mingw\lib\gcc\mingw32\5.3.0\include\c++\bits\char_traits.h:389:30: error: 'wmemset' was not decla
red in this scope
  return wmemset(__s, __a, __n);
                              ^
mingw32-make[2]: *** [FeLib/CMakeFiles/FeLib.dir/Source/bitmap.cpp.obj] Error 1
mingw32-make[1]: *** [FeLib/CMakeFiles/FeLib.dir/all] Error 2
mingw32-make: *** [all] Error 2
```

emillaine committed:
Add note about possible compile errors on MinGW with -std=c++11

This happens for me with MinGW g++ 5.3.0.

See http://stackoverflow.com/q/38538924/3425536 for more info.

emlai commented:
Added a note to `INSTALL` about enabling `-std=gnu++11` if there are compile errors.

I think this is ready to merge.

emlai commented:
@stoperro The CMake branch was finally merged to Attnam/master. Thanks a lot for helping out :)

stoperro commented:
Great :)
